server/polar/models/discount.py

```diff
@@ -57,7 +58,27 @@ class DiscountDuration(StrEnum):
 
 class Discount(MetadataMixin, RecordModel):
     __tablename__ = "discounts"
-    __table_args__ = (UniqueConstraint("organization_id", "code"),)
+
+    @declared_attr.directive
+    def __table_args__(cls) -> tuple[Index]:
+        # During tests this function is called multiple times which ends up adding the index
+        # multiple times -- leading to errors. We memoize this function to ensure we end up with
+        # the index just once.
+        if not hasattr(cls, "_memoized_indexes"):
+            _deleted_at_column = cast(
+                Column[datetime | None], cls.deleted_at
+            )  # cast to satisfy mypy
+            cls._memoized_indexes = (
+                Index(
+                    "ix_discounts_code_uniqueness",
+                    "organization_id",
+                    func.lower(cls.code),
+                    unique=True,
+                    # partial index
+                    postgresql_where=(_deleted_at_column.is_(None)),
```

Comment:
Don't really get why `cls.*` properties are not correctly typed as mapped properties but well. We could also do directly `Discount.deleted_at.is_(None)`, if you think that's better :)

Reply:
Okay, then, let's go that then 😅